pkg/didcomm/crypto/jwe/authcrypt/decrypt.go

```diff
@@ -7,54 +7,180 @@ SPDX-License-Identifier: Apache-2.0
 package authcrypt
 
 import (
+	"bytes"
+	"encoding/base64"
 	"encoding/json"
 	"errors"
+	"fmt"
 
-	"golang.org/x/crypto/blake2b"
-	"golang.org/x/crypto/chacha20poly1305"
+	"github.com/btcsuite/btcutil/base58"
+	chacha "golang.org/x/crypto/chacha20poly1305"
 	"golang.org/x/crypto/nacl/box"
 )
 
 // Decrypt will JWE decode the envelope argument for the sender and recipients
-// Using (X)Chacha20 encryption algorithm and Poly1035 authenticator
-func (c *Crypter) Decrypt(envelope []byte, recipientPrivKey *[chacha20poly1305.KeySize]byte) ([]byte, error) {
-	// TODO implement decryption and call decryptOID for the recipient's OID
+// Using (X)Chacha20 cipher and Poly1035 authenticator for the encrypted payload and
+// encrypted CEK
+// And Using (x)Salsa20 cipher with 25519 keys (libsodium equivalent) for decrypting
+// the sender's public key in the current recipient's header.
+// The current recipient is the one with the sender's encrypted key that successfully
+// decrypts with recipientPrivKey.
+func (c *Crypter) Decrypt(envelope []byte, recipientPrivKey *[chacha.KeySize]byte) ([]byte, error) {
 	jwe := &Envelope{}
 	err := json.Unmarshal(envelope, jwe)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to decrypt message: %w", err)
+	}
+	recipient, err := c.findRecipient(jwe.Recipients)
+	if err != nil || recipient == nil { // TODO check if !ErrorNotFound instead of checking recipient == nil
+		return nil, fmt.Errorf("failed to decrypt message: %w", err)
 	}
 	var oid []byte
-	for _, recipient := range jwe.Recipients {
 
-		oid, err = decryptOID(recipientPrivKey, c.sender.pub, []byte(recipient.Header.OID))
-		if oid != nil {
-			break
+	// TODO replace oid with JWK wrapped in spk recipient header
+	cryptedOID, err := base64.URLEncoding.DecodeString(recipient.Header.OID)
```

Comment:
switch to RawURLEncoding

Reply:
yes... looks like it for interoperability..

Reply:
updated